Src/Support/Google.Apis/Upload/ResumableUpload.cs

```diff
@@ -711,51 +722,45 @@ private void MediaCompleted(HttpResponseMessage response)
                     StreamLength = BytesServerReceived + LastMediaLength;
                 }
             }
-            // Set Content-Length and Content-Range.
-            int contentLength = Math.Min(ChunkSize, LastMediaLength);
-            var byteArrayContent = new ByteArrayContent(LastMediaRequest, 0, contentLength);
-            byteArrayContent.Headers.Add("Content-Range", GetContentRangeHeader(BytesServerReceived, contentLength));
-            request.Content = byteArrayContent;
-            return contentLength;
+
+            // If we've read an extra byte, it'll be included in the next chunk.
+            chunkLength = Math.Min(ChunkSize, LastMediaLength);
+            chunk = LastMediaRequest;
         }
 
         /// <summary>Prepares the given request with the next chunk in case the steam length is known.</summary>
-        private int PrepareNextChunkKnownSize(HttpRequestMessage request, Stream stream,
-            CancellationToken cancellationToken)
+        private void PrepareNextChunkKnownSize(Stream stream, CancellationToken cancellationToken, out byte[] chunk, out int chunkLength)
         {
             int chunkSize = (int)Math.Min(StreamLength - BytesServerReceived, (long)ChunkSize);
 
             // Stream length is known and it supports seek and position operations.
             // We can change the stream position and read bytes from the last point.
-            byte[] buffer = new byte[Math.Min(chunkSize, BufferSize)];
-
             // If the number of bytes received by the server isn't equal to the amount of bytes the client sent, we 
             // need to change the position of the input stream, otherwise we can continue from the current position.
             if (stream.Position != BytesServerReceived)
             {
                 stream.Position = BytesServerReceived;
             }
 
-            MemoryStream ms = new MemoryStream(chunkSize);
+            chunk = new byte[chunkSize];
+            int bytesLeft = chunkSize;
             int bytesRead = 0;
-            while (true)
+            while (bytesLeft > 0)
             {
                 cancellationToken.ThrowIfCancellationRequested();
-
-                // Read from input stream and write to output stream.
-                // TODO(peleyal): write a utility similar to (.NET 4 Stream.CopyTo method).
-                int len = stream.Read(buffer, 0, (int)Math.Min(buffer.Length, chunkSize - bytesRead));
-                if (len == 0) break;
-                ms.Write(buffer, 0, len);
+                // Make sure we only read at most BufferSize bytes at a time.
+                // TODO: Understand why this is a requirement.
```

Comment:
There may be some custom streams where the user does care about the max bytes read per read; although I would guess this is very unusual.
However, I think we should keep this behaviour as it *someone* may be relying on it, and it costs us nothing.

Reply:
Yup, that's my gut feeling too. Another example of why we should only make promises we really understand :)